clippy_lints/src/assigning_clones.rs

```diff
@@ -366,8 +367,20 @@ impl<'tcx> CallCandidate<'tcx> {
                 match self.kind {
                     CallKind::MethodCall { receiver } => {
                         let receiver_sugg = if let ExprKind::Unary(hir::UnOp::Deref, ref_expr) = lhs.kind {
-                            // `*lhs = self_expr.clone();` -> `lhs.clone_from(self_expr)`
-                            Sugg::hir_with_applicability(cx, ref_expr, "_", applicability)
+                            // If `ref_expr` implements both Deref(Mut) and Clone, we cannot remove the dereference.
+                            // Because in that case we would be calling `clone_from` on the type of `lhs` directly, but
+                            // we would also be passing it a value of type `self_expr`, which has type of `*lhs`.
+                            let ty = cx.typeck_results().expr_ty(ref_expr);
+                            let impls_clone = cx.tcx.lang_items().clone_trait().map_or(false, |id| implements_trait(cx, ty, id, &[]));
+                            let impls_deref = cx.tcx.lang_items().deref_mut_trait().map_or(false, |id| implements_trait(cx, ty, id, &[]));
+
+                            if impls_clone && impls_deref {
+                                // `*lhs = self_expr.clone();` -> `(*lhs).clone_from(self_expr)`
+                                Sugg::hir_with_applicability(cx, lhs, "_", applicability)
+                            } else {
+                                // `*lhs = self_expr.clone();` -> `lhs.clone_from(self_expr)`
+                                Sugg::hir_with_applicability(cx, ref_expr, "_", applicability)
+                            }
```

Comment:
How about only checking if `ty.is_ref()`? Same for below. They're the only safe types deref wise, it would add an explicit deref for `!Clone` wrappers but I think that's good as it avoids any future issues if the wrapper becomes `Clone`

It would also handle e.g. `*mut Waker`

Reply:
Sorry for the delay, was on vacation season. Good idea, this is probably the most conservative solution. Pushed a commit that does this

Reply:
Looks good thanks, just needs the tests blessing

Reply:
Done :)